### DISCUSSION TITLE
Better support for latex equation when publishing to confluence

### DISCUSSION THREAD

**Discussion by kascesar** [quarto-dev/quarto-cli#9659] [Feature Requests]
Hellow hope for some help :) 

Im having trouble  when publishing to confluence with this equation:

in MD mode:

```
# Introducción

$$
\text{FFT}(\psi(t))_{\boldsymbol{\delta}, \boldsymbol{\theta}, \boldsymbol{\alpha}, \boldsymbol{\beta}, \boldsymbol{\gamma}} =
\begin{cases}
\text{freqs} \in [1, 4[ (\text{hrtz}) & = \delta \\
\text{freqs} \in [4, 8[ (\text{hrtz}) & = \theta \\
\text{freqs} \in [8, 13[ (\text{hrtz}) & = \alpha \\
\text{freqs} \in [13, 30[ (\text{hrtz}) & = \beta \\
\text{freqs} \in [30, 100[ (\text{hrtz}) & = \gamma \\
\end{cases}
$$
```

im geting this error, but i dont know why:

```
[WARNING] Could not convert TeX math 
  \text{FFT}(\psi(t))_{\boldsymbol{\delta}, \boldsymbol{\theta}, \boldsymbol{\alpha}, \boldsymbol{\beta}, \boldsymbol{\gamma}} =
  \begin{cases}
  \text{freqs} \in [1, 4[ (\text{hrtz}) & = \delta \\
  \text{freqs} \in [4, 8[ (\text{hrtz}) & = \theta \\
  \text{freqs} \in [8, 13[ (\text{hrtz}) & = \alpha \\
  \text{freqs} \in [13, 30[ (\text{hrtz}) & = \beta \\
  \text{freqs} \in [30, 100[ (\text{hrtz}) & = \gamma \\
  \end{cases}
  , rendering as TeX
```

**Comment by cscheid** [comment]
This error comes from Pandoc when it cannot convert your math statement to the write format. That's what it's saying: "Could not convert TeX math". You have statements that Pandoc doesn't know how to handle. I imagine it's `\begin{cases}...\end{cases}`.

**Reply by cderv** [reply]
I don't think we do something specific for Math handling in our confluence publishing format. So probably the default engine applies: https://pandoc.org/MANUAL.html#math-rendering-in-html
> The default is to render TeX math as far as possible using Unicode characters.

Maybe we could use one of the other engine available to render Math to an image for example (using webtex to produce svg ?). 

At some point we could also come up with a feature like diagram to render math during rendering using KaTeX maybe, and producing an image. Just an idea.

I'll rename this discussion and mark as feature request

**Reply by kascesar** [reply]
> Maybe we could use one of the other engine available to render Math to an image for example (using webtex to produce svg ?).

i have a choice ? what should i do to change the render ?.

Thanks for your attention

**Reply by cderv** [reply]
This would be the configuration for HTML format: https://quarto.org/docs/output-formats/html-basics.html#latex-equations

But I don't think we support any of those for the specific Confluence publishing format which is Quarto specific. 

So for now, you can't really to much unless you do render to an image yourself using computation cell possibly or just external processing; 

Otherwise, you need to stick with not too complex equation so that the rendering to Unicode characters works. But this usally means no LaTeX command. 

So Math equation for Confluence format is not something we support yet.

**Comment by arichar6** [comment]
I'm also running into this issue, and wanted to give a +1 to this feature request! 

I tried to find the code that specifies the publishing format for Confluence, to see if there was a way that I could hack it to be able to use some of those other equation editors for the HTML format (since they render my equations just fine). But I couldn't even find the right place in the codebase to even just get started on this! 😓 

Oh well, at least for now I just wanted to agree that better equation support in the Confluence publishing would be amazing!

**Reply by cscheid** [reply]
It's not at all a trivial fix to make, so unfortunately it's not going to happen anytime soon. Confluence doesn't truly support any great math rendering, nor does it support external javascript so that we could load MathJax.

**Reply by arichar6** [reply]
That's too bad. I do wish that Confluence developers would take math rendering more seriously.

I don't suppose there's any way to use "web resources" as [discussed in this page](https://developer.atlassian.com/server/confluence/including-javascript-and-css-resources/#including-a-custom-javascript-or-css-file-from-a-plugin) to include MathJax?

**Reply by cderv** [reply]
This doc is for a Confluence Plugin development AFAIU so this won't apply. There is no easy way. 

Only thing we could "easily" do is like other format not supporting Math would be to use webtex Math option so that Math are rendered to image to be included as image in confluence. 

This is not yet possible within our Confluence format, but something we can look at as Confluence support embedding images.

**Reply by arichar6** [reply]
Thank you for looking into this!

While the webtex/image rendering would not be ideal since the images tend to look blurry, it would be much better than nothing. If it would be possible to add webtex support as an option, that would be great!

**Reply by cderv** [reply]
> While the webtex/image rendering would not be ideal since the images tend to look blurry,

Maybe confluence would support svg ? 🤔 Not sure about this.

**Reply by arichar6** [reply]
I don't know how quarto can specify cli arguments to pandoc, but maybe something like this would get the svg equations?
`pandoc --webtex 'https://latex.codecogs.com/svg.latex?' --self-contained`

https://stackoverflow.com/questions/63708025/using-pandoc-converting-a-markdown-document-with-latex-to-html-and-latex-convert

**Comment by SeanU** [comment]
Is it possible to detect what third-party macros are installed in a Confluence space, or allow users to provide a custom rendering handler that could supply their own markup to use a macro?

We use [the LaTeX macro]https://docs.adaptavist.com/cfm4cc/latest/macros/latex) from Content Formatting Macros. That hypothetically makes it easy, even if it can't be a general or default solution since it's a commercial add-on.

**Reply by cscheid** [reply]
We're not going to implement that kind of detection, because the amount of effort unfortunately doesn't justify the outcome. With that said, you're always welcome to write filters of your own to do "custom rendering" precisely: https://quarto.org/docs/extensions/filters.html#overview
